src/test/java/com/puppycrawl/tools/checkstyle/internal/ImmutabilityTest.java

```diff
@@ -0,0 +1,494 @@
+
+    /**
+     * Test to ensure that fields in util classes are immutable.
+     *
+     * @noinspection JUnitTestMethodWithNoAssertions
+     * @noinspectionreason JUnitTestMethodWithNoAssertions - asserts in callstack,
+     *     but not in this method
+     */
+    @Test
+    public void testUtilClassesImmutability() {
+        final JavaClasses utilClasses = new ClassFileImporter()
+            .withImportOption(ImportOption.Predefined.DO_NOT_INCLUDE_TESTS)
+            .importPackages("com.puppycrawl.tools.checkstyle.utils",
+                            "com.puppycrawl.tools.checkstyle.checks.javadoc.utils");
+
+        final ArchCondition<JavaField> beSuppressedField = new SuppressionArchCondition<>(
+            SUPPRESSED_FIELDS_IN_UTIL_CLASSES, "be suppressed");
+
+        final ArchRule fieldsShouldBeImmutable = fields()
+            .that()
+            .areDeclaredInClassesThat()
+            .haveSimpleNameEndingWith("Util")
+            .should(BE_IMMUTABLE)
+            .andShould()
+            .beFinal()
+            .andShould()
+            .beStatic()
+            .orShould(beSuppressedField);
+
+        fieldsShouldBeImmutable.check(utilClasses);
+    }
+
+    /**
+     * Test to ensure modules annotated with {@link StatelessCheck} contain immutable fields.
+     *
+     * @noinspection JUnitTestMethodWithNoAssertions
+     * @noinspectionreason JUnitTestMethodWithNoAssertions - asserts in callstack,
+     *     but not in this method
+     */
+    @Test
+    public void testFieldsInStatelessChecksShouldBeImmutable() {
+        final DescribedPredicate<JavaField> moduleProperties = new ModulePropertyPredicate();
+
+        final ArchCondition<JavaField> beSuppressedField = new SuppressionArchCondition<>(
+            SUPPRESSED_FIELDS_IN_MODULES, "be suppressed");
+
+        final ArchCondition<JavaField> bePrivateExceptMessageKeys =
+            new ArchCondition<>("be private") {
+
+                @Override
+                public void check(JavaField item, ConditionEvents events) {
+                    if (!item.getName().startsWith("MSG_")
+                        && !item.getModifiers().contains(JavaModifier.PRIVATE)) {
+                        final String message = String
+                            .format(Locale.ROOT, "Field <%s> should %s in %s",
+                                    item.getFullName(), getDescription(),
+                                    item.getSourceCodeLocation().toString());
+                        events.add(SimpleConditionEvent.violated(item, message));
+                    }
+                }
+            };
+
+        final ArchRule fieldsInStatelessChecksShouldBeImmutable = fields()
+            .that()
+            .areDeclaredInClassesThat()
+            .areAnnotatedWith(StatelessCheck.class)
+            .and(are(not(moduleProperties)))
+            .should(BE_IMMUTABLE)
+            .andShould()
+            .beFinal()
+            .andShould(bePrivateExceptMessageKeys)
```

Comment:
I think I am getting cross-eyed here, but all fields should be immutable period regardless of public/private etc... for Stateless.

There is a discrepancy here. Test is called `testFieldsInStatelessChecksShouldBeImmutable`. It does not mention the type of fields (private), so one would naturally assume there is no type requirement.

It seems to me `andShould(bePrivateExceptMessageKeys)` so should be a separate test but apply to all modules as there should be no direct communication between isolated checks.

Reply:
Immutability condition is being equally applied to all fields, it is just that `MSG_` are allowed to be private but other conditions before that are still applicable to them, it is just they are allowed to be public, if a `MSG_` isn't final, we will get a test failure.

Reply:
> if a MSG_ isn't final

We only uppercase fields that are static/final. So all MSGs have to be final, otherwise they would be `msgXxx`.

I also don't see how this applies to my comments. Private should be separate archunit as immutability doesn't care about modifier.

Reply:
Done, removed public, private part.

Reply:
@Vyom-Yadav This is done, but are we moving the `private` test to a new PR for ArchUnit?

Reply:
We can, feel free to edit https://github.com/checkstyle/checkstyle/issues/12099#issuecomment-1233749876, so that we can have a clearer goal.

Reply:
Added what I understand the `private` check to be as number 6 in the comment recommended.